File: backtest/engine.py
import pandas as pd
import numpy as np
from .costs import calculate_costs


class BacktestEngine:
    def __init__(self, initial_capital=10000, commission=0.001, slippage=0.0005):
        self.initial_capital = initial_capital
        self.commission = commission
        self.slippage = slippage

    def run(self, data, strategy, train_pct=0.7):
        df = data.copy()
        n = len(df)
        train_end = int(n * train_pct)

        train_data = df.iloc[:train_end]
        test_data = df.iloc[train_end:]

        # generate signals on train, evaluate on test
        train_signals = strategy.generate_signals(train_data)
        test_signals = strategy.generate_signals(test_data)

        train_results = self._simulate(train_signals)
        test_results = self._simulate(test_signals)

        return {
            'train': train_results,
            'test': test_results,
            'strategy': strategy.name,
            'params': strategy.get_params()
        }

    def _simulate(self, data):
        df = data.copy()
        df = df.dropna()

        capital = self.initial_capital
        position = 0
        trades = []
        equity_curve = []

        for i, row in df.iterrows():
            price = row['close']
            signal = row['signal']

            if signal == 1 and position == 0:
                # buy
                shares = capital / price
                cost = calculate_costs(capital, self.commission, self.slippage)
                capital -= cost
                position = shares
                trades.append({'type': 'buy', 'price': price, 'shares': shares, 'date': row.get('date', i)})

            elif signal == -1 and position > 0:
                # sell
                trade_value = position * price
                cost = calculate_costs(trade_value, self.commission, self.slippage)
                capital = trade_value - cost
                trades.append({'type': 'sell', 'price': price, 'shares': position, 'date': row.get('date', i)})
                position = 0

            # track equity
            equity = capital if position == 0 else position * price
            equity_curve.append({'date': row.get('date', i), 'equity': equity, 'price': price})

        equity_df = pd.DataFrame(equity_curve)
        final_equity = equity_curve[-1]['equity'] if equity_curve else self.initial_capital

        return {
            'equity_curve': equity_df,
            'trades': trades,
            'final_equity': final_equity,
            'return_pct': (final_equity - self.initial_capital) / self.initial_capital * 100,
            'num_trades': len(trades)
        }


# TODO: add short selling support
# TODO: position sizing options

File: lib/backtest/engine.ts
/**
 * Backtest Engine
 * Core simulation engine for trading strategies
 */

import { Strategy, SignalData, OHLCV } from '../strategies/types';
import { calculateCosts } from './costs';
import {
  BacktestResult,
  SimulationResult,
  Trade,
  EquityPoint,
  CostModel,
} from './types';

export interface EngineConfig {
  initialCapital?: number;
  commission?: number;
  slippage?: number;
  costModel?: CostModel;
}

export class BacktestEngine {
  private initialCapital: number;
  private commission: number;
  private slippage: number;
  private costModel: CostModel;

  constructor(config: EngineConfig = {}) {
    this.initialCapital = config.initialCapital ?? 10000;
    this.commission = config.commission ?? 0.001;
    this.slippage = config.slippage ?? 0.0005;
    this.costModel = config.costModel ?? 'fixed';
  }

  /**
   * Run backtest with train/test split
   */
  run(data: OHLCV[], strategy: Strategy, trainPct: number = 0.7): BacktestResult {
    const n = data.length;
    const trainEnd = Math.floor(n * trainPct);

    const trainData = data.slice(0, trainEnd);
    const testData = data.slice(trainEnd);

    // Generate signals separately for train and test
    const trainSignals = strategy.generateSignals(trainData);
    const testSignals = strategy.generateSignals(testData);

    // Simulate both periods
    const trainResults = this.simulate(trainSignals);
    const testResults = this.simulate(testSignals);

    return {
      train: trainResults,
      test: testResults,
      strategy: strategy.name,
      params: strategy.getParams(),
    };
  }

  /**
   * Simulate trading on signal data
   */
  simulate(signals: SignalData[]): SimulationResult {
    // Filter out any signals with NaN values
    const data = signals.filter(
      (s) => !Number.isNaN(s.close) && s.signal !== undefined
    );

    let capital = this.initialCapital;
    let position = 0;
    const trades: Trade[] = [];
    const equityCurve: EquityPoint[] = [];

    for (const row of data) {
      const price = row.close;
      const signal = row.signal;

      // Buy signal and no position
      if (signal === 1 && position === 0) {
        const shares = capital / price;
        const cost = calculateCosts(capital, {
          commissionPct: this.commission,
          slippagePct: this.slippage,
          model: this.costModel,
          price,
        });
        capital -= cost;
        position = shares;
        trades.push({
          type: 'buy',
          price,
          shares,
          date: row.date,
          value: capital,
        });
      }
      // Sell signal and have position
      else if (signal === -1 && position > 0) {
        const tradeValue = position * price;
        const cost = calculateCosts(tradeValue, {
          commissionPct: this.commission,
          slippagePct: this.slippage,
          model: this.costModel,
          price,
        });
        capital = tradeValue - cost;
        trades.push({
          type: 'sell',
          price,
          shares: position,
          date: row.date,
          value: capital,
        });
        position = 0;
      }

      // Track equity
      const equity = position === 0 ? capital : position * price;
      equityCurve.push({
        date: row.date,
        equity,
        price,
      });
    }

    const finalEquity =
      equityCurve.length > 0
        ? equityCurve[equityCurve.length - 1].equity
        : this.initialCapital;

    return {
      equityCurve,
      trades,
      finalEquity,
      returnPct: ((finalEquity - this.initialCapital) / this.initialCapital) * 100,
      numTrades: trades.length,
    };
  }

  /**
   * Get engine configuration
   */
  getConfig(): EngineConfig {
    return {
      initialCapital: this.initialCapital,
      commission: this.commission,
      slippage: this.slippage,
      costModel: this.costModel,
    };
  }
}
